### PR TITLE
7915 checks in l2arc_evict could use some cleaning up

### PR DESCRIPTION
**Pull request by avg-I**
https://www.illumos.org/issues/7915

### CODE REVIEW

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by avg-I** [comment]
In `l2arc_evict` for any buffer visited on `l2ad_buflist` the following must be true:
- it has an l2 header
- it's not being written to a cache device
- it's not a write head

Another minor change is that the current code can evict a buffer that starts (on disk) right after the requested eviction range.  There is no need to do that.

**Comment by avg-I** [comment]
@grwilson ping

**Comment by avg-I** [comment]
Looks like zloop failure is unrelated to the change.

**Comment by avg-I** [comment]
@skiselkov could you please take a look as well?

**Comment by avg-I** [comment]
Build fails with
```
../../common/fs/zfs/blkptr.c:144: error: passing argument 2 of 'zio_decompress_data' from incompatible pointer type
dmake: Warning: Command failed for target `obj64/blkptr.o'
```
This change does not come anywhere close to the failing code.

**Comment by yuripv** [comment]
It's https://www.illumos.org/issues/8267 (not related to your change).

**Comment by rmustacc** [comment]
@avg-l the problem is that the recent merge of 8067 zdb should be able to dump literal embedded block pointer, broke the build. You're seeing collateral damage.

**Comment by prakashsurya** [comment]
@avg-I sorry for the trouble; the build failure is my fault.

**Comment by avg-I** [comment]
Thank you all for the information and thank you for fixing the problem.

File: usr/src/uts/common/fs/zfs/arc.c [hunk 1]
```diff
@@ -6709,18 +6709,16 @@ l2arc_evict(l2arc_dev_t *dev, uint64_t distance, boolean_t all)
...
+		 */
+		ASSERT(HDR_HAS_L2HDR(hdr));
 
-		if (!all && HDR_HAS_L2HDR(hdr) &&
-		    (hdr->b_l2hdr.b_daddr > taddr ||
+		/* Ensure this header has finished being written. */
+		ASSERT(!HDR_L2_WRITING(hdr));
+		ASSERT(!HDR_L2_WRITE_HEAD(hdr));
+
+		if (!all && (hdr->b_l2hdr.b_daddr >= taddr ||
```
Comment (dankimmel) [inline usr/src/uts/common/fs/zfs/arc.c]: Did you mean to change this `>` to `>=` here? (If so could you quickly clarify why?)

Comment (avg-I) [inline usr/src/uts/common/fs/zfs/arc.c]: Yes, I did. `taddr` is the end of what we should evict (start + size). So, if a buffer starts at `taddr`, then it's beyond the area to be evicted.  That one buffer, if any, does not make a lot of difference, but just for the sake of correctness.

**Comment by prakashsurya** [comment]
@dankimmel we can ignore the "pr-merge" result; I can give you more details offline if you're interested. the "pr-head" run was successful, so this looks good w.r.t. the automated build and tests.

**Comment by dankimmel** [comment]
Great, sounds good 👍

**Comment by avg-I** [comment]
Is there anything else that I should do for this PR?

**Comment by prakashsurya** [comment]
Thanks for your patience @avg-I. This LGTM; I'll rebase this to get an up-to-date build/test run, and open the RTI today.

**Comment by avg-I** [comment]
@prakashsurya, thank you very much.  I think that the change should apply cleanly to the latest master.

**Pull request closed without merging**